### ISSUE TITLE
ArgumentError: undefined class/module Struct::SetSendingMessagesToSentForSchoolJob

### ISSUE THREAD

**Issue by hoenth** [salsify/delayed_job_groups_plugin#7]
Thank you very much for this plugin/gem. It was just what I needed to have a job run once all other related jobs had completed. 

I am trying use the on_completion_job feature. Here is how I have it setup

      job_group = Delayed::JobGroups::JobGroup.create!(on_completion_job: SetSendingMessagesToSentForSchoolJob.new(school_id),
                                                      on_completion_job_options: { queue: 'default' })

      job_group.enqueue(Newsletters::SendNewsletterToParentsJob.new(school_id: @school.id, year: school.current_year), queue: 'default')
      job_group.enqueue(Newsletters::SendNewsletterToStudentsJob.new(school_id: @school.id, year: school.current_year), queue: 'default')
      job_group.enqueue(Newsletters::SendNewsletterToStaffMembersJob.new(school_id: @school.id, year: school.current_year), queue: 'default')
      job_group.enqueue(Newsletters::SendNewsletterToTeachersJob.new(school_id: @school.id, year: school.current_year), queue: 'default')

      job_group.mark_queueing_complete

However, when it comes time to run the SetSendingMessagesToSentForSchoolJob, I get the following error:

    ArgumentError: undefined class/module Struct::SetSendingMessagesToSentForSchoolJob
    /Users/thomashoen/.rvm/gems/ruby-2.2.3@cp_4/gems/activerecord-4.2.7.1/lib/active_record/coders/yaml_column.rb:26:in `load'
    /Users/thomashoen/.rvm/gems/ruby-2.2.3@cp_4/gems/activerecord-4.2.7.1/lib/active_record/type/serialized.rb:19:in `type_cast_from_database'
    /Users/thomashoen/.rvm/gems/ruby-2.2.3@cp_4/gems/activerecord-4.2.7.1/lib/active_record/attribute.rb:104:in `type_cast'
    /Users/thomashoen/.rvm/gems/ruby-2.2.3@cp_4/gems/activerecord-4.2.7.1/lib/active_record/attribute.rb:42:in `original_value'
    /Users/thomashoen/.rvm/gems/ruby-2.2.3@cp_4/gems/activerecord-4.2.7.1/lib/active_record/attribute.rb:37:in `value'
    /Users/thomashoen/.rvm/gems/ruby-2.2.3@cp_4/gems/activerecord-4.2.7.1/lib/active_record/attribute_set.rb:31:in `fetch_value'
    /Users/thomashoen/.rvm/gems/ruby-2.2.3@cp_4/gems/activerecord-4.2.7.1/lib/active_record/attribute_methods/read.rb:93:in `_read_attribute'
    /Users/thomashoen/.rvm/gems/ruby-2.2.3@cp_4/gems/activerecord-4.2.7.1/lib/active_record/attribute_methods.rb:50:in `__temp__f6e6f536f6d607c6564796f6e6f5a6f626'
    /Users/thomashoen/.rvm/gems/ruby-2.2.3@cp_4/gems/delayed_job_groups_plugin-0.3.0/lib/delayed/job_groups/job_group.rb:94:in `complete'
    /Users/thomashoen/.rvm/gems/ruby-2.2.3@cp_4/gems/delayed_job_groups_plugin-0.3.0/lib/delayed/job_groups/job_group.rb:77:in `block in check_for_completion'

I inserted a pry statement at that point, and received the same error when I attempted to view the `on_completion_job` value. I could, however, manually run `SetSendingMessagesToSentForSchoolJob.new(25)` while in that pry session. 

Am I setting the on_completion_job incorrectly?

**Comment by jturkel** [comment]
It looks like you are setting the `on_completion_job` correctly and there's some kind of problem with classloading. Is the `SetSendingMessagesToSentForSchoolJob` class nested inside another class or module? If so, have you tried using a fully qualified class name?

**Comment by hoenth** [comment]
@jturkel Thanks for taking a look at this. Originally, the class was under a module `Newsletters`, and I included the namespace when setting the `on_complete_job`, but I got the same error. So I thought that it could be an issue with namespacing, so moved the class out of the namespace. 

What I don't under

I have also tried initializing the class in the on_complete_job call like this:
    Delayed::JobGroups::JobGroup.create!(on_completion_job: ::SetSendingMessagesToSentForSchoolJob.new(school_id),
                                                  on_completion_job_options: { queue: 'default' })
but got the same result

Here is what happens in a debugging session:

    [90, 99] in /Users/thomashoen/.rvm/gems/ruby-2.2.3@cp_4/gems/delayed_job_groups_plugin-0.3.0/lib/delayed/job_groups/job_group.rb
       90:         queueing_complete? && !JobGroup.has_pending_jobs?(id) && !blocked?
       91:       end
       92: 
       93:       def complete
       94:         byebug
    => 95:         Delayed::Job.enqueue(on_completion_job, on_completion_job_options || {}) if on_completion_job
       96:         destroy
       97:       end
       98:     end
       99:   end
    (byebug) on_completion_job.to_s
    *** ArgumentError Exception: undefined class/module SetSendingMessagesToSentForSchoolJob

    nil
    (byebug) SetSendingMessagesToSentForSchoolJob
    SetSendingMessagesToSentForSchoolJob
    (byebug) on_completion_job.to_s
    "#<SetSendingMessagesToSentForSchoolJob:0x007f90f8bc0e90>"
    (byebug) on_completion_job
    #<SetSendingMessagesToSentForSchoolJob:0x007f90f8bc0e90 @school_id=25>

As you can see, when I first attempt to access the `on_completion_job` variable, I get the error. But if I type the name of the class into the console, there is no error. And after that, referring to the variable no longer causes an error. 

The class is in the app/jobs folder. 

Any idea why the class is not being loaded when trying to call the variable the first time?

**Comment by hoenth** [comment]
I added `job_group.enqueue(SetSendingMessagesToSentForSchoolJob.new(school_id), queue: 'default')` and removed the on_completion_job callbacks. All jobs queued and ran as expected. 

After that test, I reverted back to setting the on_completion_job call and reran the parent job that enqueues the jobs, but this time I did not stop and start the DJ job workers. Since I had previously loaded the class, by manually entering the name of the class in a pry console, the "complete" method successfully queued the call. So perhaps this is only an issue with running DJ in development.

**Comment by hoenth** [comment]
I have tried a couple other tests. If I enqueue the class, as I did in the comment above, then I can also set the on_completion_job and I don't get errors. But if the on_completion_job references a class that has not been seen/used/referenced before, it fails. 

But, if a reference the class in one of the other jobs that are enqueued (I inserted a `SetSendingMessagesToSentForSchoolJob.new(0)` in one of the other job's perform method, I don't get the error. 

This (https://alisdair.mcdiarmid.org/fixing-rails-auto-loading-for-serialized-objects/), I assume, is related to the issue. On de-serialization, the class is not being reloaded. So this is a development-only issue. Closing, as I have a work around.
